tests/test_state/test_storage_state.py

```diff
@@ -33,3 +34,111 @@ def test_market_cycle():
             assert storage_state.offered_buy_kWh[future_time_slots[0]] == 0
             assert storage_state.offered_sell_kWh[future_time_slots[0]] == 0
+
+    @staticmethod
+    def _setup_storage_state_for_clamp_energy():
+        storage_state = StorageState()
+        current_time_slot = now()
+        market_slot_list = [current_time_slot,
+                            current_time_slot + duration(minutes=15),
+                            current_time_slot + duration(minutes=30),
+                            current_time_slot + duration(minutes=45)]
+        storage_state._current_market_slot = current_time_slot
+        storage_state._used_storage = 250.0
+        storage_state.capacity = 500.0
+        for time_slot in market_slot_list:
+            storage_state.offered_sell_kWh[time_slot] = 0.0
+            storage_state.pledged_sell_kWh[time_slot] = 0.0
+            storage_state.offered_buy_kWh[time_slot] = 0.0
+            storage_state.pledged_buy_kWh[time_slot] = 0.0
+        return storage_state, market_slot_list
+
+    def test_clamp_energy_to_sell_kWh_only_on_first_slot(self):
+        storage_state, market_slot_list = self._setup_storage_state_for_clamp_energy()
+
+        # Enable battery to sell all its available capacity in one market slot
+        storage_state._battery_energy_per_slot = storage_state.capacity
+        storage_state._clamp_energy_to_sell_kWh(market_slot_list)
+        expected_available_energy = (
+                storage_state.used_storage -
+                storage_state.min_allowed_soc_ratio * storage_state.capacity)
+        expected_energy_to_sell = {
+            market_slot_list[0]: expected_available_energy,
+            market_slot_list[1]: 0.0,
+            market_slot_list[2]: 0.0,
+            market_slot_list[3]: 0.0,
+        }
+        assert expected_energy_to_sell == storage_state.energy_to_sell_dict
+
+    def test_clamp_energy_to_sell_kWh_respects_battery_energy_per_slot(self):
+        storage_state, market_slot_list = self._setup_storage_state_for_clamp_energy()
+
+        storage_state._battery_energy_per_slot = 25.0
+        storage_state._clamp_energy_to_sell_kWh(market_slot_list)
+        expected_energy_to_sell = {
+            market_slot_list[0]: 25.0,
+            market_slot_list[1]: 25.0,
+            market_slot_list[2]: 25.0,
+            market_slot_list[3]: 25.0,
+        }
+        assert expected_energy_to_sell == storage_state.energy_to_sell_dict
+
+    def test_clamp_energy_to_sell_kWh_respects_energy_offers_trades(self):
+        storage_state, market_slot_list = self._setup_storage_state_for_clamp_energy()
+
+        storage_state._battery_energy_per_slot = 25.0
+        for time_slot in market_slot_list:
+            storage_state.offered_sell_kWh[time_slot] = 5.0
+            storage_state.pledged_sell_kWh[time_slot] = 15.0
+        storage_state._clamp_energy_to_sell_kWh(market_slot_list)
```

Comment:
I understand that it would make no sense to offer/promise more energy than it is able to deliver, and therefore the quantities you use on your tests respect `_battery_energy_per_slot >= offered_sell_kWh + pledged_sell_kWh` condition. 
Therefore, I think of a test in which this is not like that. For example this same test but with pledges and offers of 15.0 kWh each. In this case `storage_dict[time_slot] = self._max_offer_energy_kWh(time_slot) = -5`, and afterwards, the available_energy_for_all_slots would end up growing after `available_energy_for_all_slots -= -5`, and make it less likely to be assigned to `energy_to_sell[future_time_slot]` on future time_slots. Also, energy_to_sell_dict would have several negative values, which do have an interpretation on _there is some energy to bid for instead of sell/offer_ , however being this is the purpose of energy_to_buy_dict, I'm not sure this is the right behavior. 
Of course this scenario is hypothetical and would only be enabled by setting conditions which break this `_battery_energy_per_slot >= offered_sell_kWh + pledged_sell_kWh`.

However, I wonder how is this scenario avoided? In other words, how can we be sure that `_battery_energy_per_slot >= offered_sell_kWh + pledged_sell_kWh` always happens?

Reply:
@ERafaelMartinez to answer your last question, the scenario that you describe is avoided only by the calling code unfortunately. The way this class works from the outside (to be more precise, from the trading strategies) , is that the caller code first has to get the available energy to sell / buy (using the get_available_energy_to_buy/sell_kWh methods) which takes into account the battery energy per slot in order for the condition to not be violated, post the offer / bid with the available energy returned by the method, and then updates the pledged energy by calling register_energy_from_posted_offer/bid. That way it is guaranteed that the amount of energy that is placed as an offer or bid never violates this condition. 
I predict that on the next refactoring cycle this will be one of the most important things to refactor in this class, in order to ensure that this condition is never violated from inside this class. However for now I would go with Hannes' recommendation on the ticket comments, and add assertions, in order to at least get notified that this condition will never be violated. 
I'll also try to add unit tests that validate the assertion, I know that they will probably be deprecated soon, however I will add them in case the aforementioned refactoring is not done soon.

Reply:
Thank you for explaining, and sorry for the nitpickiness

Reply:
> Thank you for explaining, and sorry for the nitpickiness

Please don't be thanks again for catching that cornercase !